File: Editor/src/main/java/com/rspsi/studio/WorkspaceManager.java
package com.rspsi.studio;

import com.rspsi.cache.workspace.CacheSessionState;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Tracks which workspaces are open as persistent tabs and which one is focused.
 *
 * <p>Opening a workspace that is already open only changes focus; it never implies
 * reloading that workspace's content. Callers use {@link #isOpen(Workspace)} before
 * doing expensive (re)load work, and {@link #close(Workspace)} to actually dispose one.</p>
 */
public final class WorkspaceManager {
    public enum Workspace {
        DASHBOARD,
        MAP_EDITOR,
        INTERFACE_STUDIO,
        OBJECT_STUDIO
    }

    private final LinkedHashSet<Workspace> open = new LinkedHashSet<>(Set.of(Workspace.DASHBOARD));
    private Workspace active = Workspace.DASHBOARD;
    private Workspace pendingFocus;

    public Workspace active() {
        return active;
    }

    /** Open workspaces (tabs), in the order they were first opened. Dashboard is always present. */
    public Set<Workspace> open() {
        return Collections.unmodifiableSet(open);
    }

    public boolean isOpen(Workspace workspace) {
        return open.contains(workspace);
    }

    public void openDashboard() {
        focus(Workspace.DASHBOARD);
    }

    public boolean openMapEditor(CacheSessionState cacheState) {
        if (cacheState != CacheSessionState.READY) return false;
        open.add(Workspace.MAP_EDITOR);
        focus(Workspace.MAP_EDITOR);
        return true;
    }

    public boolean openInterfaceStudio(CacheSessionState cacheState) {
        if (cacheState != CacheSessionState.READY) return false;
        open.add(Workspace.INTERFACE_STUDIO);
        focus(Workspace.INTERFACE_STUDIO);
        return true;
    }

    public boolean openObjectStudio(CacheSessionState cacheState) {
        if (cacheState != CacheSessionState.READY) return false;
        open.add(Workspace.OBJECT_STUDIO);
        focus(Workspace.OBJECT_STUDIO);
        return true;
    }

    /** Removes a tab. Dashboard can never be closed; it is the permanent home tab. */
    public void close(Workspace workspace) {
        if (workspace == Workspace.DASHBOARD) return;
        open.remove(workspace);
        if (active == workspace) {
            focus(Workspace.DASHBOARD);
        }
    }

    public void focus(Workspace workspace) {
        active = workspace;
        pendingFocus = workspace;
    }

    /** Consumes the pending focus request; used to force-select a tab in the native tab bar once. */
    public Workspace consumePendingFocus() {
        Workspace value = pendingFocus;
        pendingFocus = null;
        return value;
    }

    /** Reconciles focus with whichever tab the native tab bar reports as selected this frame. */
    public void setActiveFromTabBar(Workspace workspace) {
        active = workspace;
    }
}

File: Editor/src/main/kotlin/com/rspsi/studio/WorkspaceManager.kt
package com.rspsi.studio

import com.rspsi.cache.workspace.CacheSessionState
import java.util.Collections
import java.util.LinkedHashSet

/**
 * Tracks which workspaces are open as persistent tabs and which one is focused.
 *
 * Opening a workspace that is already open only changes focus; it never implies
 * reloading that workspace's content. Callers use [isOpen] before doing expensive
 * (re)load work, and [close] to actually dispose one.
 */
class WorkspaceManager {
    enum class Workspace {
        DASHBOARD,
        MAP_EDITOR,
        INTERFACE_STUDIO,
        OBJECT_STUDIO,
    }

    private val openWorkspaces = LinkedHashSet<Workspace>().apply {
        add(Workspace.DASHBOARD)
    }

    private var activeWorkspace = Workspace.DASHBOARD
    private var pendingFocus: Workspace? = null

    fun active(): Workspace = activeWorkspace

    /** Open workspaces (tabs), in the order first opened. Dashboard is always present. */
    fun open(): Set<Workspace> = Collections.unmodifiableSet(openWorkspaces)

    fun isOpen(workspace: Workspace): Boolean = workspace in openWorkspaces

    fun openDashboard() {
        focus(Workspace.DASHBOARD)
    }

    fun openMapEditor(cacheState: CacheSessionState): Boolean =
        openWhenReady(cacheState, Workspace.MAP_EDITOR)

    fun openInterfaceStudio(cacheState: CacheSessionState): Boolean =
        openWhenReady(cacheState, Workspace.INTERFACE_STUDIO)

    fun openObjectStudio(cacheState: CacheSessionState): Boolean =
        openWhenReady(cacheState, Workspace.OBJECT_STUDIO)

    /** Resets project workspaces when switching or closing the active Studio project. */
    fun reset() {
        openWorkspaces.clear()
        openWorkspaces.add(Workspace.DASHBOARD)
        activeWorkspace = Workspace.DASHBOARD
        pendingFocus = Workspace.DASHBOARD
    }

    /** Removes a tab. Dashboard can never be closed; it is the permanent home tab. */
    fun close(workspace: Workspace) {
        if (workspace == Workspace.DASHBOARD) return

        openWorkspaces.remove(workspace)
        if (activeWorkspace == workspace) {
            focus(Workspace.DASHBOARD)
        }
    }

    fun focus(workspace: Workspace) {
        activeWorkspace = workspace
        pendingFocus = workspace
    }

    /** Consumes the pending focus request used to force-select a native tab exactly once. */
    fun consumePendingFocus(): Workspace? =
        pendingFocus.also {
            pendingFocus = null
        }

    /** Reconciles focus with whichever tab the native tab bar reports as selected this frame. */
    fun setActiveFromTabBar(workspace: Workspace) {
        activeWorkspace = workspace
    }

    private fun openWhenReady(
        cacheState: CacheSessionState,
        workspace: Workspace,
    ): Boolean {
        if (cacheState != CacheSessionState.READY) return false

        openWorkspaces.add(workspace)
        focus(workspace)
        return true
    }
}
